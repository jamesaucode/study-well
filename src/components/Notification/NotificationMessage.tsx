import React from "react";
import styled from "styled-components";
import { Message } from "../../styles/shared";
import { NextFC } from "next";
import { IMessage } from "../../../interfaces/message";
import SVG from "react-inlinesvg";

interface WrapperProps {
  mounted?: boolean;
}
const Wrapper = styled.div<WrapperProps>`
  display: flex;
  justify-content: center;
  border-radius: 5px;
  width: 100%;
  padding: 0.75rem;
  box-sizing: border-box;
  transition: 0.5s ease-in-out opacity;
`;
const Important = styled.span`
  font-weight: 700;
`;
const SVGWrapper = styled.div`
  padding: 0 0.5rem;
`;
const renderImportant = (text: string): React.ReactElement => (
  <Important>{text}</Important>
);
const NotificationMessage: NextFC<IMessage> = ({
  success,
  message,
  id,
  // delay = 3000,
  removeNotification,
}): JSX.Element => {
  return (
    <Wrapper key={id}>
      <Message success={success}>
        <div>
          {success ? renderImportant("Success! ") : renderImportant("Oops! ")}
          {message}
        </div>
        <SVGWrapper onClick={(): void => removeNotification(id)}>
          <SVG className="small-icon" src="/static/images/close.svg" />
        </SVGWrapper>
      </Message>
    </Wrapper>
  );
};

NotificationMessage.defaultProps = {
  success: false,
  message: "No message",
};

export default React.memo(NotificationMessage);
