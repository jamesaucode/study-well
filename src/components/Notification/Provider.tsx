import React, { useRef, useState, useEffect } from 'react';
import { Context } from './Context';

const Provider = () => {
    const root  = useRef<Node | null>(null);
    const id = useRef([]);
    const [notifications, setNotifications] = useState([]);

    useEffect(() => {
        root.current = document.createElement('div');
        document.body.appendChild(root.current);
        return () => {
           id.current.forEach(clearTimeout);
           if (root.current) {
            document.body.removeChild(root.current);
           }
        };
    }, [])

    // const remove = (notification) => {

    // }
}
