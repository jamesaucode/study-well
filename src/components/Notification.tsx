import React, { useState, useEffect, useRef } from "react";
import NotificationMessage from "./NotificationMessage";
import { MessageType } from "../../typings/message";

const Notification = (props: any) => {
    const [notifications, setNotifications] = useState();
    const popNotification = () => {
        setNotifications(notifications.slice(0, -1));
    }
    const pushNotification = (message  = "lmaooooo", status: boolean) => {
        setNotifications(notifications.concat({
            message,
            success: status
        }))
    }

  const children = React.Children.map(props.children, child => {
      return React.cloneElement(child, {
          popNotification,
          pushNotification 
      })
  })
  if (notifications) {
    return (
      <React.Fragment>
        {notifications.map((notification: MessageType) => {
          return (
            <NotificationMessage
              message={notification.message}
              success={notification.success}
            />
          );
        })}
        {children}
      </React.Fragment>
    );
  } else {
    return <button onClick={popNotification}>Test Button</button>;
  }
};

export default React.memo(Notification);
