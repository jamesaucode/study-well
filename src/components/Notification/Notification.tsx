import React, { useState, useEffect, useRef, EffectCallback } from "react";
import NotificationMessage from "./NotificationMessage";
import { IMessage } from "../../../interfaces/message";
import styled from "styled-components";
import { useWindowSize } from "../../hooks/useWindowSize";
import uuid from "uuid";

interface PropTypes {
  children: React.ReactChild;
}

interface AllNotificationWrapperProps {
  windowHeight: number;
  os: number;
}

interface NotificationContextProps {
  pushNotification: (message: string, success: boolean) => void;
  removeNotification: (message: string, success: boolean) => void;
}
const AllNotificationsWrapper = styled.div<AllNotificationWrapperProps>`
  height: 0;
  top: 0px;
  font-size: calc(0.35vw + 16px);
  width: 100%;
  position: absolute;
  z-index: 1000;
`;
const NotificationContext = React.createContext<NotificationContextProps | null>(
  null,
);
export const NotificationProvider: React.FunctionComponent<PropTypes> = ({
  children,
}): JSX.Element => {
  const [notifications, setNotifications] = useState<IMessage[]>([]);
  const windowSize = useWindowSize();
  const delay = 5000;
  const id = useRef<number[]>([]);
  useEffect((): EffectCallback => {
    return (): void => {
      id.current.forEach(clearTimeout);
    };
  }, []);
  const removeNotification = (notificationId: string): void => {
    setNotifications((prevState): IMessage[] => {
      return prevState.filter((n): boolean => n.id !== notificationId);
    });
  };
  const pushNotification = (message = "", status: boolean): IMessage => {
    const notification = {
      message,
      success: status,
      id: uuid(),
      delay,
    };
    const timerId = setTimeout((): void => {
      removeNotification(notification.id);
      id.current.splice(id.current.indexOf(timerId), 1);
    }, delay);
    id.current.push(timerId);
    setNotifications((allNotification): IMessage[] =>
      allNotification.concat(notification),
    );
    return notification;
  };

  const value = {
    pushNotification,
    removeNotification,
  };
  if (notifications) {
    return (
      <NotificationContext.Provider value={value}>
        <AllNotificationsWrapper
          windowHeight={windowSize.size.windowHeight}
          os={notifications.length * 54}
        >
          {notifications.map(
            (notification: IMessage): React.ReactElement => {
              return (
                <NotificationMessage
                  key={notification.id}
                  {...notification}
                  removeNotification={removeNotification}
                />
              );
            },
          )}
        </AllNotificationsWrapper>
        {children}
      </NotificationContext.Provider>
    );
  } else {
    return <React.Fragment>{children}</React.Fragment>;
  }
};

export const useNotification = (): NotificationContextProps =>
  React.useContext(NotificationContext);
