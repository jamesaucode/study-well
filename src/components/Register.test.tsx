/* eslint-disable */
import { mount } from "enzyme";
import Register from "./Register";
import { NotificationProvider } from "./Notification/Notification";

describe("Register", () => {
  const fakeOnClick = jest.fn(() => {});
  it("Component renders", () => {
    const wrapper = mount(
      <NotificationProvider>
        <Register toggleForm={fakeOnClick} />
      </NotificationProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
