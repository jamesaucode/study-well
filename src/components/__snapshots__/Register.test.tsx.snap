// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component renders 1`] = `
<NotificationProvider>
  <Notification__AllNotificationsWrapper
    os={0}
    windowHeight={768}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "Notification__AllNotificationsWrapper-sc-14z8rhj-0",
            "isStatic": true,
            "lastClassName": "kdjGqQ",
            "rules": Array [
              "height:0;top:0px;font-size:calc(0.35vw + 16px);width:100%;position:absolute;z-index:1000;",
            ],
          },
          "displayName": "Notification__AllNotificationsWrapper",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "Notification__AllNotificationsWrapper-sc-14z8rhj-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      os={0}
      windowHeight={768}
    >
      <div
        className="Notification__AllNotificationsWrapper-sc-14z8rhj-0 kdjGqQ"
      />
    </StyledComponent>
  </Notification__AllNotificationsWrapper>
  <Register
    toggleForm={[MockFunction]}
  >
    <Register__FormWrapper>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "Register__FormWrapper-sc-1b4znjf-2",
              "isStatic": true,
              "lastClassName": "jLdrlS",
              "rules": Array [
                "max-width:",
                "400px",
                ";font-size:calc(0.35vw + 16px);height:fit-content;text-align:center;",
              ],
            },
            "displayName": "Register__FormWrapper",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "Register__FormWrapper-sc-1b4znjf-2",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="Register__FormWrapper-sc-1b4znjf-2 jLdrlS"
        >
          <Register__FormInput
            name="email"
            onChange={[Function]}
            placeholder="Email"
            required={true}
            type="email"
            value=""
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "Register__FormInput-sc-1b4znjf-1",
                    "isStatic": true,
                    "lastClassName": "iYJDBT",
                    "rules": Array [
                      "border:1px solid #aaaaaa;border-radius:2px;padding:0.75rem;margin:0.5rem 0;width:100%;box-sizing:border-box;font-size:0.7em;&:focus{border:1px solid #8610f9;}",
                    ],
                  },
                  "displayName": "Register__FormInput",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "Register__FormInput-sc-1b4znjf-1",
                  "target": "input",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              name="email"
              onChange={[Function]}
              placeholder="Email"
              required={true}
              type="email"
              value=""
            >
              <input
                className="Register__FormInput-sc-1b4znjf-1 iYJDBT"
                name="email"
                onChange={[Function]}
                placeholder="Email"
                required={true}
                type="email"
                value=""
              />
            </StyledComponent>
          </Register__FormInput>
          <Register__FormInput
            name="password"
            onChange={[Function]}
            placeholder="Password"
            required={true}
            type="password"
            value=""
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "Register__FormInput-sc-1b4znjf-1",
                    "isStatic": true,
                    "lastClassName": "iYJDBT",
                    "rules": Array [
                      "border:1px solid #aaaaaa;border-radius:2px;padding:0.75rem;margin:0.5rem 0;width:100%;box-sizing:border-box;font-size:0.7em;&:focus{border:1px solid #8610f9;}",
                    ],
                  },
                  "displayName": "Register__FormInput",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "Register__FormInput-sc-1b4znjf-1",
                  "target": "input",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              name="password"
              onChange={[Function]}
              placeholder="Password"
              required={true}
              type="password"
              value=""
            >
              <input
                className="Register__FormInput-sc-1b4znjf-1 iYJDBT"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                required={true}
                type="password"
                value=""
              />
            </StyledComponent>
          </Register__FormInput>
          <Register__FormInput
            name="firstName"
            onChange={[Function]}
            placeholder="First Name"
            required={true}
            type="text"
            value=""
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "Register__FormInput-sc-1b4znjf-1",
                    "isStatic": true,
                    "lastClassName": "iYJDBT",
                    "rules": Array [
                      "border:1px solid #aaaaaa;border-radius:2px;padding:0.75rem;margin:0.5rem 0;width:100%;box-sizing:border-box;font-size:0.7em;&:focus{border:1px solid #8610f9;}",
                    ],
                  },
                  "displayName": "Register__FormInput",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "Register__FormInput-sc-1b4znjf-1",
                  "target": "input",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              name="firstName"
              onChange={[Function]}
              placeholder="First Name"
              required={true}
              type="text"
              value=""
            >
              <input
                className="Register__FormInput-sc-1b4znjf-1 iYJDBT"
                name="firstName"
                onChange={[Function]}
                placeholder="First Name"
                required={true}
                type="text"
                value=""
              />
            </StyledComponent>
          </Register__FormInput>
          <Register__FormInput
            name="lastName"
            onChange={[Function]}
            placeholder="Last Name"
            required={true}
            type="text"
            value=""
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "Register__FormInput-sc-1b4znjf-1",
                    "isStatic": true,
                    "lastClassName": "iYJDBT",
                    "rules": Array [
                      "border:1px solid #aaaaaa;border-radius:2px;padding:0.75rem;margin:0.5rem 0;width:100%;box-sizing:border-box;font-size:0.7em;&:focus{border:1px solid #8610f9;}",
                    ],
                  },
                  "displayName": "Register__FormInput",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "Register__FormInput-sc-1b4znjf-1",
                  "target": "input",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              name="lastName"
              onChange={[Function]}
              placeholder="Last Name"
              required={true}
              type="text"
              value=""
            >
              <input
                className="Register__FormInput-sc-1b4znjf-1 iYJDBT"
                name="lastName"
                onChange={[Function]}
                placeholder="Last Name"
                required={true}
                type="text"
                value=""
              />
            </StyledComponent>
          </Register__FormInput>
          <Register__FormSubmit
            onClick={[Function]}
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "Register__FormSubmit-sc-1b4znjf-3",
                    "isStatic": true,
                    "lastClassName": "sUdqg",
                    "rules": Array [
                      "color:white;font-size:0.75em;font-weight:600;width:100%;border:none;border-radius:3px;padding:0.65rem;background:#8610f9;&:hover{cursor:pointer;}",
                    ],
                  },
                  "displayName": "Register__FormSubmit",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "Register__FormSubmit-sc-1b4znjf-3",
                  "target": "button",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              onClick={[Function]}
            >
              <button
                className="Register__FormSubmit-sc-1b4znjf-3 sUdqg"
                onClick={[Function]}
              >
                Sign Up!
              </button>
            </StyledComponent>
          </Register__FormSubmit>
          <style__FormBottom>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "style__FormBottom-oz4l5n-1",
                    "isStatic": false,
                    "lastClassName": "fNxKQZ",
                    "rules": Array [
                      "font-size:0.8em;padding:1rem 0;color:",
                      [Function],
                      ";",
                    ],
                  },
                  "displayName": "style__FormBottom",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "style__FormBottom-oz4l5n-1",
                  "target": "p",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <p
                className="style__FormBottom-oz4l5n-1 fNxKQZ"
              >
                New user ? 
                <Register__StyledLink
                  onClick={[MockFunction]}
                >
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "Register__StyledLink-sc-1b4znjf-0",
                          "isStatic": true,
                          "lastClassName": "cHuoKM",
                          "rules": Array [
                            "color:blue;text-decoration:underline;cursor:pointer;font-weight:bold;",
                          ],
                        },
                        "displayName": "Register__StyledLink",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "Register__StyledLink-sc-1b4znjf-0",
                        "target": "a",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onClick={[MockFunction]}
                  >
                    <a
                      className="Register__StyledLink-sc-1b4znjf-0 cHuoKM"
                      onClick={[MockFunction]}
                    >
                      Login 
                    </a>
                  </StyledComponent>
                </Register__StyledLink>
                here!
              </p>
            </StyledComponent>
          </style__FormBottom>
        </div>
      </StyledComponent>
    </Register__FormWrapper>
  </Register>
</NotificationProvider>
`;
